### PR TITLE
AT and NVS integration 

### LINKED ISSUE #126: Add communication between wifi-subsys and the main MCU using UART port
#### Description

Add the communication between modulo de wifi and uart the handler AT is in module uart this will be used for to execute functions of the wifi for we can controlled since it.

### PR DESCRIPTION
**Pull request by xkevin190**
### Contribution description

Was added the modification of the module wifi for be controlled since uart.

### Testing procedure

cd test && idf.py build -D "TEST_COMPONENTS=network" flash monitor

```c

Press ENTER to see the list of tests.
I (2803) wifi:ap channel adjust o:6,1 n:4,1
I (2803) wifi:new:<4,0>, old:<6,1>, ap:<4,1>, sta:<4,0>, prof:6
I (3633) wifi:state: init -> auth (b0)
I (3643) wifi:state: auth -> assoc (0)
I (3643) wifi:state: assoc -> run (10)
I (3673) wifi:connected with Chicho, aid = 26, channel 4, BW20, bssid = 60:12:3c:21:03:88
I (3673) wifi:security: WPA2-PSK, phy: bgn, rssi: -64
I (3683) wifi:pm start, type: 1

I (3723) wifi:AP's beacon interval = 102400 us, DTIM period = 1
I (6683) esp_netif_handlers: sta ip: 172.16.0.159, mask: 255.255.255.0, gw: 172.16.0.1
I (6683) Wifi Module: got ip:172.16.0.159

Here's the test menu, pick your combo:
(1)	"set default credentials" [network]
(2)	"Wi-Fi initialize" [network]
(3)	"change ssid ap" [network]
(4)	"change password ap" [network]
(5)	"change ssid sta" [network]
(6)	"change password ap" [network]

```

for test uart functions.

cd test && idf.py build -D "TEST_COMPONENTS=uart" flash monitor

### Issues/PRs references

this pr fixes #126 

This PR still is in progress but you already see the code and share with me your opinions.

### CODE REVIEW

File: wifi-subsys/components/network/src/wifi.h [hunk 1]
```diff
@@ -89,6 +89,76 @@ esp_err_t set_default_credentials();
...
+/**
+ * @brief this function set the wifi mode and execute the connection,
+ *         it find the wifi mode in the nvs and it verify if already saved
+ *         the mode in the nvs. If the nvs return error or the wifi_mode is
+ *         equal to WIFI_MODE_NULL will take the defaults params for give 
+ *         warranty to the user that wifi init correctly this is very important
+ *         because if something failed  the wifi  will  stop.
+ * 
+ */
+esp_err_t set_wifi_mode();
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.h:157]: `set_wifi_mode` and `change_wifi_mode` sound similar,  I'm thinking  if both could be condensed in a single function

Comment (luisan00) [reply wifi-subsys/components/network/src/wifi.h:157]: Also as parameter the related `ESP_WIFI_MODE` enum could be used

**Review by luisan00 — COMMENTED** [review]
Anyway this job is looking really good 👍

File: wifi-subsys/components/network/src/wifi.h (hunk 1 above)
Comment (xkevin190) [reply wifi-subsys/components/network/src/wifi.h:157]: sounds similar but are two deferents  functions  `set_wifi_mode` is an reusable funtcions used many times insde in the Wifi module,  `change wifi mode` just receive the uart command and set new mode in the nvs and restart the wifi, maybe i should change for a better name.

Comment (luisan00) [reply wifi-subsys/components/network/src/wifi.h:157]: so one it's only for internal use and the other one as AT command :+1:

File: wifi-subsys/.gitignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
 sdkconfig
 sdkconfig.old
 dev-config.h
+managed_components/
+dependencies.lock
```
Comment (eduazocar) [inline wifi-subsys/.gitignore:5]: add a new EOF here

**Comment by xkevin190** [comment]
work flow completed, with unit test in uart and wifi.

File: wifi-subsys/components/network/src/wifi.c [hunk 3]
```diff
@@ -76,10 +76,33 @@ static void event_handler(void *arg, esp_event_base_t event_base,
...
+    }
+
+    err = nvs_get_string(stringlify(WSTA),
+                        stringlify(WSTA_PASS),
+                        &sta_config.sta.password,
+                        &password_len);
+
+
+    ESP_LOGI(TAG, "ssid sta is %s", sta_config.sta.ssid);
+    ESP_LOGI(TAG, "password sta is %s", sta_config.sta.password);
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:101]: it's better in  a single  line

File: wifi-subsys/components/network/src/wifi.c [hunk 4]
```diff
@@ -201,6 +277,31 @@ esp_err_t set_default_credentials(void)
...
+    }
+    err = nvs_set_uint8(stringlify(WIFI),
+                        stringlify(WIFI_MODE),
+                        identify_enabled_interface());
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "Error: to setting WIFI mode to the nvs");
+    }
     /* cppcheck-suppress duplicateCondition
     * (reason: <your reason why you think this is a false positive>) */
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:306]: reason?

File: wifi-subsys/components/network/src/wifi.c [hunk 5]
```diff
@@ -217,9 +318,262 @@ esp_err_t set_default_credentials(void)
     return ESP_OK;
 }
 
-esp_err_t wifi_init(void)
+esp_err_t wifi_on(void)
+{
+  esp_err_t err =  wifi_init();
+  if (err != ESP_OK) {
+      return err;
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:325]: LOG error

File: wifi-subsys/components/network/src/wifi.c [hunk 6]
```diff
@@ -217,9 +318,262 @@ esp_err_t set_default_credentials(void)
...
+   }
+   return ESP_OK;
+}
+
+esp_err_t wifi_restart(void)
+{
+    esp_err_t err;
+    err = wifi_off();
+    if(err != ESP_OK){
+        return err;
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:345]: LOG error or debug

File: wifi-subsys/components/network/src/wifi.c [hunk 7]
```diff
@@ -217,9 +318,262 @@ esp_err_t set_default_credentials(void)
...
+    }
+    err = wifi_on();
+    if(err != ESP_OK){
+        return err;
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:349]: same here, log it

File: wifi-subsys/components/network/src/wifi.c [hunk 8]
```diff
@@ -217,9 +318,262 @@ esp_err_t set_default_credentials(void)
...
+    esp_err_t err;
+    err = nvs_set_string(stringlify(WSTA), stringlify(WSTA_SSID), sta_ssid);
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "Error: to settin STA ssid to the nvs");
+    }
+    else {
+        err = init_sta();
+        if (err != ESP_OK) {
+           return err;
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:366]: log it

File: wifi-subsys/components/network/src/wifi.c [hunk 9]
```diff
@@ -217,9 +318,262 @@ esp_err_t set_default_credentials(void)
...
+    ESP_LOGE(TAG,"esp_set_string %s", ap_ssid);
+    err = nvs_set_string(stringlify(WAP), stringlify(WAP_SSID), ap_ssid);
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "Error: to settin STA ssid to the nvs");
+    }
+    else {
+       err = init_ap();
+       if (err != ESP_OK) {
+           return err;
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:384]: log it

File: wifi-subsys/components/network/src/wifi.c [hunk 10]
```diff
@@ -217,9 +318,262 @@ esp_err_t set_default_credentials(void)
...
+
+    esp_err_t err = nvs_set_uint8(stringlify(WAP), stringlify(WAP_AUTH),auth);
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "Error: to setting AP auth to the nvs");
+    }
+
+    err = set_wifi_mode();
+    if (err != ESP_OK) {
+        return err;
```
Comment (luisan00) [inline wifi-subsys/components/network/src/wifi.c:484]: log it

**Comment by xkevin190** [comment]
this pr is ready for review, waiting for a new review.

**Comment by luisan00** [comment]
Looks good for me, some request from @RocioRojas or @eduazocar?

File: wifi-subsys/components/network/test/test_wifi.c [hunk 11]
```diff
@@ -16,6 +16,16 @@
 
 #include "unity.h"
 #include "wifi.h"
+#include "storage.h"
+
+TEST_CASE("set default credentials", "[network]")
+{
+   nvs_init();
+   esp_err_t err = set_default_credentials();
```
Comment (eduazocar) [inline wifi-subsys/components/network/test/test_wifi.c:24]: Indent here is 3, the used is 4

File: wifi-subsys/test/dependencies.lock [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+      type: service
+    version: 0.5.4
+  idf:
+    component_hash: null
+    source:
+      type: idf
+    version: 5.0.0
+manifest_hash: cbf68eaa6a45e0d806821655af9f819fc8a95f52334076795faa4e4ee899c87a
+target: esp32
+version: 1.0.0
```
Comment (eduazocar) [inline wifi-subsys/test/dependencies.lock:15]: This file is required? or to be ignored?. I am confuse with the file .gitignore that refers to this file..

**Review by eduazocar — CHANGES_REQUESTED** [review]

File: wifi-subsys/test/dependencies.lock (hunk 12 above)
Comment (luisan00) [reply wifi-subsys/test/dependencies.lock:15]: this should be ignored i think

**Review by luisan00 — APPROVED** [review]
ready!

**Review by eduazocar — APPROVED** [review]

**Pull request merged**